### ISSUE TITLE
New install problem: Could not find the metadata of an IdP with entity ID

### ISSUE THREAD

**Issue by Beedell** [catalyst/moodle-auth_saml2#393]
I have recently installed version 2019110701 of auth_saml2 installed on my Moodle 3.7 development server, run upgrade, turned on auth_saml2 and added these configurations in the plugin:

- idpmetadata - added as xml returned by the idp metadata url (because adding this url directly is rejected by the form validation) https://fidm.eu1.gigya.com/saml/v2.0/3_RkYTZbeIWPjABvIlrfXcT3msogGyMurlyfhKlQFNVJ2YHmPJZDksTAWBicQqHFNb/idp/metadata
- debug - yes
- logtofile - yes (with a suitable path added as well)

I have also regenerated the certificate (just in case), downloaded my SP metadata and sent to the IdP admin who has added it to the site.

In a fresh browser when I go to [dirroot]/auth/saml2/test.php to test authentication, and click on 'login' I get this:

> Exception - Could not find the metadata of an IdP with entity ID '5afdd3fd43484647cca8b580cb45dcf4'
Stack trace:
    line 298 of /auth/saml2/extlib/simplesamlphp/modules/saml/lib/Auth/Source/SP.php: SimpleSAML\Error\Exception thrown
    line 692 of /auth/saml2/extlib/simplesamlphp/modules/saml/lib/Auth/Source/SP.php: call to SimpleSAML\Module\saml\Auth\Source\SP->getIdPMetadata()
    line 795 of /auth/saml2/extlib/simplesamlphp/modules/saml/lib/Auth/Source/SP.php: call to SimpleSAML\Module\saml\Auth\Source\SP->startSSO()
    line 193 of /auth/saml2/extlib/simplesamlphp/lib/SimpleSAML/Auth/Source.php: call to SimpleSAML\Module\saml\Auth\Source\SP->authenticate()
    line 165 of /auth/saml2/extlib/simplesamlphp/lib/SimpleSAML/Auth/Simple.php: call to SimpleSAML\Auth\Source->initLogin()
    line 108 of /auth/saml2/extlib/simplesamlphp/lib/SimpleSAML/Auth/Simple.php: call to SimpleSAML\Auth\Simple->login()
    line 72 of /auth/saml2/test.php: call to SimpleSAML\Auth\Simple->requireAuth()

I have tried the 'fixes' mentioned in these previous issues #338, #322, #36, and have tried switching to another IdP (https://samltest.id/) but get the same exception.

So I have looked into what is going wrong in the code on my install. The error message tells us that the $entityId is an md5 encoded version of the IdP metadata url. auth/saml2/extlib/simplesamlphp/lib/SimpleSAML/Metadata/MetaDataStorageSource.php lookupIndexFromEntityId() expects the $entityId to be a plain text version of the IdP metadata url.

A fix for this error is to add this to line 265:
`            if (md5($index) === $entityId) {`
`                return $index;`
`            }`

Now with this fix I can get to the IdP login page.

But after logging in I get a second error:

> SAML2 exception: Cannot retrieve metadata for IdP 'https://fidm.gigya.com/saml/v2.0/3_mDxR4kcZ6X3IwdrDVmyMuxLhGbeVPL7-15Z5BINRdJIaxGBAaz8gRqP4_6yZ7Ku0' because it isn't a valid IdP for this SP.
Debug info: #0 [dirroot]/auth/saml2/extlib/simplesamlphp/modules/saml/www/sp/saml2-acs.php(130): SimpleSAML\Module\saml\Auth\Source\SP->getIdPMetadata('https://fidm.gi...')
#1 [dirroot]/auth/saml2/sp/saml2-acs.php(32): require('/srv/jb23347.vl...')

This time we see that the $idp is the plain text version of the IdP metadata url and not the md5 encoded version. So the fix is in line 54-57 of auth/saml2/extlib/simplesamlphp/modules/saml/www/sp/saml2-acs.php:

`$idp = md5($issuer);`
`if ($issuer instanceof \SAML2\XML\saml\Issuer) {`
`    $idp = md5($idp->getValue());`
`}`

I hope you can reproduce these errors. They are rather fundamental to the success of your plugin. If it is some configuration or setup issue then I would be very grateful to know. Alternatively these fixes appear to work on my install, though I have not checked unit/behat tests (there may be much better fixes).

**Comment by brendanheywood** [comment]
@Beedell sorry about that, I can actually reproduce this with the test page, with both your idp and my test one, but my test one actually works fine for real logins. So I think this is a bug in the test.php page itself which I'm digging into now.

So does your idp work ok for a proper full moodle login?

**Comment by brendanheywood** [comment]
@Beedell can you please retest using the latest master branch and see how you go?

**Comment by Beedell** [comment]
Wow, what a service!

I only submitted an issue just before going home last night, and first thing in the morning I find a really good improvement to the test.php page. I really like the extra information you have added to this page.

Many, many thanks Brendan. 

I can confirm both the test.php page and normal SAML2 login are both working for me now with your changes in the master branch. Closing this issue as fixed.
